File: .cursor/skills/sleuths/sleuth/src/query.rs
use crate::config::sleuths_dir;
use anyhow::{Context, Result};
use std::path::Path;

#[derive(Debug, serde::Deserialize)]
pub struct SleuthQuery {
    pub id: String,
    pub description: String,
    pub prompt: String,
}

pub fn load_query(project_root: &Path, sleuth_id: &str) -> Result<SleuthQuery> {
    let path = sleuths_dir(project_root)
        .join("queries")
        .join(format!("{sleuth_id}.yaml"));
    let raw = std::fs::read_to_string(&path)
        .with_context(|| format!("read sleuth query {}", path.display()))?;
    let mut q: SleuthQuery = serde_yaml::from_str(&raw)?;
    if q.id.is_empty() {
        q.id = sleuth_id.to_string();
    }
    Ok(q)
}

pub fn list_sleuth_ids(project_root: &Path) -> Result<Vec<String>> {
    let dir = sleuths_dir(project_root).join("queries");
    if !dir.is_dir() {
        return Ok(vec![]);
    }
    let mut ids = Vec::new();
    for entry in std::fs::read_dir(&dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) == Some("yaml") {
            if let Some(stem) = path.file_stem().and_then(|s| s.to_str()) {
                ids.push(stem.to_string());
            }
        }
    }
    ids.sort();
    Ok(ids)
}

File: .cursor/skills/sleuths/sleuth/query.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import yaml

from sleuth.config import sleuths_dir


@dataclass
class SleuthQuery:
    id: str
    description: str
    prompt: str


def load_query(project_root: Path, sleuth_id: str) -> SleuthQuery:
    path = sleuths_dir(project_root) / "queries" / f"{sleuth_id}.yaml"
    with path.open(encoding="utf-8") as f:
        raw = yaml.safe_load(f) or {}
    qid = raw.get("id") or sleuth_id
    return SleuthQuery(
        id=str(qid),
        description=str(raw["description"]),
        prompt=str(raw["prompt"]),
    )


def list_sleuth_ids(project_root: Path) -> list[str]:
    queries_dir = sleuths_dir(project_root) / "queries"
    if not queries_dir.is_dir():
        return []
    ids = [
        p.stem
        for p in queries_dir.iterdir()
        if p.is_file() and p.suffix == ".yaml"
    ]
    ids.sort()
    return ids
